src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1220,6 +1229,15 @@
      :state (conj state (su/munge-jsonb-for-storage
                           (path->nested-map path value)))}))
 
+(defn parse-json-regex-query
```

Comment:
This is the param extractor for a JsonbRegexExpression. At a high level these things zip through the query plan, pulling parameters out of the plan into the state vector, and replacing them with placeholders (e.g '?').

In this case, the input looks like

```
[{:field "parameters.message.foo" :value "bar"} ["a" "b" "c"]]
```

and the output looks like

```
[{:field "parameters" :value ["?" "?" "?"]} ["a" "b" "c" "message" "foo" "bar" "message"]]
```

that field and value are eventually passed to su/jsonb-regex, which turns them into proper SQL. Does that clarify things any? Would it help if I renamed it to jsonb-param-extractor?